pl-faded-parsons.py

```diff
@@ -1,305 +1,577 @@
+            self.pre_text or self.post_text
+        ):
+            raise Exception(
+                "pre-text and post-text are not supported in right (horizontal) mode. "
+                + 'Add/set `format="bottom"` or `format="no-code"` to your element to use this feature.'
+            )
+
+        path = pl.get_string_attrib(
+            element, "solution-path", "./solution"
+        )
+        self._solution_path = os.path.join(data["options"]["question_path"], path)
+        self._max_distractors = 10  # this was hardcoded before
+        self._raw_answers = data["raw_submitted_answers"]
+        self._options = data["options"]
+
+        # load the trays and log fields
+        if self.answers_name in self._raw_answers:
+            prev_submission: Submission = cast(
+                Submission,
+                validate_and_instantiate(
+                    Submission, self._raw_answers[self.answers_name]
+                ),
+            )
+            self._trays_from_submission(prev_submission)
+        else:
+            self._trays_from_markup()
+
+    def _trays_from_markup(self) -> None:
+        starters, givens, distractors = [], [], []
+        BLANK = re.compile(r"#blank [^#]*")
+        GIVEN = re.compile(r"#(\d+)given")
+        DISTRACTOR = re.compile(r"#distractor")
+
+        for raw_line in self.markup.strip().split("\n"):
+            line_str = raw_line.strip()
+            snippets = line_str.split("#")[0].split("!BLANK")
+
+            blanks = [""] * (len(snippets) - 1)
+            if blank_defaults := re.findall(BLANK, line_str):
+                for i, val in enumerate(blank_defaults):
+                    blanks[i] = val.replace("#blank", "").strip()
+
+            if match := re.search(GIVEN, line_str):
+                givens.append(Submission.Line(int(match.group(1)), snippets, blanks))
+            else:
+                line = Submission.Line(0, snippets, blanks)
+                if re.search(DISTRACTOR, line_str):
+                    distractors.append(line)
+                else:
+                    starters.append(line)
+
+        distractor_count = min(len(distractors), self._max_distractors)
+        starters.extend(random.sample(distractors, k=distractor_count))
+
+        random.shuffle(starters)
+
+        self.trays: Submission.Trays
+        if self.format == FadedParsonsProblem.Format.NO_CODE:
+            self.trays = Submission.Trays(solution=givens + starters, starter=[])
+        else:
+            self.trays = Submission.Trays(solution=givens, starter=starters)
+        self.log: List[Submission.LogEntry] = []
 
-def render_submission_panel(element, data):
-    """Show student what they submitted"""
-    html_params = {
-        'code': get_student_code(element, data),
-    }
-    with open('pl-faded-parsons-submission.mustache', 'r') as f:
-        return chevron.render(f, html_params).strip()
+    def _trays_from_submission(self, data: Submission) -> None:
```

Comment:
Something is going wrong here, the data isn't being restored...

Reply:
I believe this is fixed by [28a926e](https://github.com/ace-lab/pl-faded-parsons/pull/51/commits/28a926e1a235dc93724ceffd63ed75249397d8a1)